### PR TITLE
build ood components from source code for aarch64

### LINKED ISSUE #116: OnDemand aarch64 builds
We added multi-arch builds to support M1 macs. All seems to work fine except for OnDemand as only x86_64 rpm builds exist. Is it possible to get aarch64 builds? If not, what is arch specific? We now install TurboVNC directly from their yum repo which provides aarch64 builds. We also install python-websockify from source. Same for Dex, we just build the binary directly. What else is there?

### PR DESCRIPTION
**Pull request by johrstrom**
Fixes #116 by building ood components from source code for aarch64.

I tested it as much as I could (by hardcoding `ARCHTYPE`) but ran into #119. Which is to say it boots up and seems to work as well as Slurm will let it.

### CODE REVIEW

**Comment by johrstrom** [comment]
@aebruno don't know if you saw this. I don't have an M1, but I hardcoded ARCHTYPE to test it.

**Comment by aebruno** [comment]
> @aebruno don't know if you saw this. I don't have an M1, but I hardcoded ARCHTYPE to test it.

Thanks @johrstrom. Testing this out now, will report back if I hit any issues.

File: ondemand/install.sh [hunk 1]
```diff
@@ -7,6 +7,183 @@ log_info() {
...
+  git clone https://github.com/zmartzone/mod_auth_openidc.git
+  cd mod_auth_openidc
+  git checkout v2.4.9.4
+  sh autogen.sh
+  ./configure
+  make && make install
+
+  cd $BUILD_DIR
+  git clone https://github.com/OSC/ondemand.git
+  cd ondemand
```
Comment (johrstrom) [inline ondemand/install.sh:151]: I'm now remembering that I probably need to set this back to a tag instead the mainline.

Comment (aebruno) [reply ondemand/install.sh:151]: Yeah good idea. Should be an easy fix.

File: ondemand/install.sh [hunk 2]
```diff
@@ -7,6 +7,185 @@ log_info() {
...
+  cd "$BUILD_DIR/passenger-$PASS_VERSION"
+  tar xf $BUILD_DIR/nginx.tar.gz
+
+  PREFIX=/opt/ood/ondemand/root
+  NGINX_DATADIR=$PREFIX/usr/share/nginx
+  NGINX_CONFDIR=$PREFIX/etc/nginx
+  NGINX_HOME=/var/lib/ondemand-nginx
+  NGINX_HOME_TMP=$NGINX_HOME/tmp
+  NGINX_LOGDIR=/var/log/ondemand-nginx
+  BASE_CCOPTS='-g -pipe -Wall -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector-strong --param=ssp-buffer-size=4 -grecord-gcc-switches -m64 -mtune=generic'
```
Comment (aebruno) [inline ondemand/install.sh:28]: Let's replace these options `-m64 -mtune=generic` with the following `-march=native -mtune=native`:

```diff
-  BASE_CCOPTS='-g -pipe -Wall -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector-strong --param=ssp-buffer-size=4 -grecord-gcc-switches -m64 -mtune=generic'
+  BASE_CCOPTS='-g -pipe -Wall -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector-strong --param=ssp-buffer-size=4 -grecord-gcc-switches -march=native -mtune=native'
```

File: ondemand/install.sh [hunk 3]
```diff
@@ -7,6 +7,185 @@ log_info() {
...
+  NGINX_CCOPTS="-O2 $BASE_CCOPTS"
+  PASSENGER_CCOPTS="$BASE_CCOPTS -Wno-deprecated"
+  LDOPTS="-Wl,-z,relro -Wl,-E"
+  RUBY_LIBDIR=$PREFIX/usr/share/ruby/vendor_ruby
+
+  rake nginx OPTIMIZE=yes CACHING=false
+  cd "$BUILD_DIR/passenger-$PASS_VERSION/nginx-$NGINX_VERSION"
+  ./configure \
```
Comment (aebruno) [inline ondemand/install.sh:36]: The arm64 builds failed unless all these config opts were on a single line. There must be some whitespace issues going on here but placing them all on a single line fixed the issues for the arm64 builds:

```diff
-  ./configure \
-         --prefix=$NGINX_DATADIR \
-         --sbin-path=$PREFIX/usr/sbin/nginx \
-    --conf-path=$NGINX_CONFDIR/nginx.conf \
-    --error-log-path=$NGINX_LOGDIR/error.log \
-    --http-log-path=$NGINX_LOGDIR/access.log \
-    --http-client-body-temp-path=$NGINX_HOME_TMP/client_body \
-    --http-proxy-temp-path=$NGINX_HOME_TMP/proxy \
-    --http-fastcgi-temp-path=$NGINX_HOME_TMP/fastcgi \
-    --http-uwsgi-temp-path=$NGINX_HOME_TMP/uwsgi \
-    --http-scgi-temp-path=$NGINX_HOME_TMP/scgi \-    --pid-path=/run/ondemand-nginx.pid \
-    --lock-path=/run/lock/subsys/ondemand-nginx \
-    --user=ondemand-nginx \
-    --group=ondemand-nginx \
-    --with-file-aio \
-    --with-http_ssl_module \-    --with-http_v2_module \
-    --with-http_realip_module \
-    --with-http_addition_module \
-    --with-http_xslt_module \-    --with-http_image_filter_module \-    --with-http_sub_module \
-    --with-http_dav_module \
-    --with-http_flv_module \
-    --with-http_mp4_module \
-    --with-http_gunzip_module \
-    --with-http_gzip_static_module \
-    --with-http_random_index_module \
-    --with-http_secure_link_module \
-    --with-http_degradation_module \
-    --with-http_stub_status_module \
-    --with-mail \
-    --with-mail_ssl_module \
-    --with-pcre \-    --with-pcre-jit \
-    --add-module=../src/nginx_module \
-    --with-cc-opt="$NGINX_CCOPTS" \-    --with-ld-opt="$LDOPTS" \-    --with-debug
+  ./configure --prefix=$NGINX_DATADIR --sbin-path=$PREFIX/usr/sbin/nginx --conf-path=$NGINX_CONFDIR/nginx.conf --error-log-path=$NGINX_LOGDIR/error.log --http-log-path=$NGINX_LOGDIR/access.log --http-client-body-temp-path=$NGINX_HOME_TMP/client_body --http-proxy-temp-path=$NGINX_HOME_TMP/proxy --http-fastcgi-temp-path=$NGINX_HOME_TMP/fastcgi --http-uwsgi-temp-path=$NGINX_HOME_TMP/uwsgi --http-scgi-temp-path=$NGINX_HOME_TMP/scgi --pid-path=/run/ondemand-nginx.pid --lock-path=/run/lock/subsys/ondemand-nginx --user=ondemand-nginx --group=ondemand-nginx --with-file-aio --with-http_ssl_module --with-http_v2_module --with-http_realip_module --with-http_addition_module --with-http_xslt_module --with-http_image_filter_module --with-http_sub_module --with-http_dav_module --with-http_flv_module --with-http_mp4_module --with-http_gunzip_module --with-http_gzip_static_module --with-http_random_index_module --with-http_secure_link_module --with-http_degradation_module --with-http_stub_status_module --with-mail --with-mail_ssl_module --with-pcre --with-pcre-jit --add-module=../src/nginx_module --with-cc-opt="$NGINX_CCOPTS" --with-ld-opt="$LDOPTS" --with-debug
```

File: ondemand/install.sh [hunk 4]
```diff
@@ -7,6 +7,185 @@ log_info() {
...
+  mkdir -p  $NGINX_DATADIR/html
+  mkdir -p  $NGINX_CONFDIR
+  mkdir -p  $NGINX_HOME
+  mkdir -p  $NGINX_HOME_TMP
+  mkdir -p  $NGINX_LOGDIR
+
+  set -x
+  cd "$BUILD_DIR/passenger-$PASS_VERSION"
+  which ruby
+	rake fakeroot \
```
Comment (aebruno) [inline ondemand/install.sh:87]: Had a similar whitepsace issue here. All these env vars needed to be on a single line or else the arm64 builds failed:

```diff
-       rake fakeroot \
-           NATIVE_PACKAGING_METHOD=rpm \
-           FS_PREFIX=$PREFIX \
-           FS_BINDIR=$PREFIX/bin \
-           FS_SBINDIR=$PREFIX/sbin \
-           FS_DATADIR=$PREFIX/usr/share \
-           FS_LIBDIR=$PREFIX/lib64 \
-           FS_DOCDIR=$PREFIX/usr/share/doc \
-           RUBY=$(which ruby) \
-           RUBYLIBDIR=$RUBY_LIBDIR \
-           RUBYARCHDIR=$RUBY_LIBDIR \
-           APACHE2_MODULE_PATH=$PREFIX/usr/lib/apache2/modules/mod_passenger.so \
-                       OPTIMIZE=yes \
-                       CACHING=false \
-                       EXTRA_CFLAGS="$PASSENGER_CCOPTS" \
-                       EXTRA_CXXFLAGS="$PASSENGER_CCOPTS"
+  rake fakeroot NATIVE_PACKAGING_METHOD=rpm FS_PREFIX=$PREFIX FS_BINDIR=$PREFIX/bin FS_SBINDIR=$PREFIX/sbin FS_DATADIR=$PREFIX/usr/share FS_LIBDIR=$PREFIX/lib64 FS_DOCDIR=$PREFIX/usr/share/doc RUBY=$(which ruby) RUBYLIBDIR=$RUBY_LIBDIR RUBYARCHDIR=$RUBY_LIBDIR APACHE2_MODULE_PATH=$PREFIX/usr/lib/apache2/modules/mod_passenger.so OPTIMIZE=yes CACHING=false EXTRA_CFLAGS="$PASSENGER_CCOPTS" EXTRA_CXXFLAGS="$PASSENGER_CCOPTS"
```

File: ondemand/install.sh [hunk 5]
```diff
@@ -7,6 +7,185 @@ log_info() {
...
+  dnf -y update && \
+    dnf install -y dnf-utils epel-release && \
+    dnf config-manager --set-enabled powertools && \
+    dnf -y module enable nodejs:12 ruby:2.7 && \
+    dnf install -y \
+        file lsof sudo gcc gcc-c++ git \
+        patch lua-posix rsync ruby ruby-devel python2 python3 \
+        nodejs sqlite sqlite-devel nmap-ncat httpd httpd-devel mod_ssl \
+        libcurl-devel autoconf openssl-devel jansson-devel libxml2-devel \
+        libxslt-devel gd-devel
```
Comment (aebruno) [inline ondemand/install.sh:125]: Needed to add `libaio-devel` package here.

File: ondemand/install.sh [hunk 6]
```diff
@@ -7,6 +7,185 @@ log_info() {
...
+  sh autogen.sh
+  ./configure
+  make && make install
+
+  cd $BUILD_DIR
+  OOD_VERSION='2.0.27'
+  wget "https://github.com/OSC/ondemand/archive/refs/tags/v$OOD_VERSION.tar.gz"
+  tar -xf "v$OOD_VERSION.tar.gz"
+  cd ondemand-$OOD_VERSION
+  bundle config --local path ~/vendor/bundle
```
Comment (aebruno) [inline ondemand/install.sh:154]: Looks like Nokogiri gem includes pre-compiled binaries by default. The binaries built for arm64 require `GLIBC_2.29` which is not in rockylinux8. So we need to add these 2 lines here to force Nokogiri to build from source:

```diff
+  bundle config build.nokogiri --use-system-libraries
+  bundle config set force_ruby_platform true
```

For reference, here's the error I was getting: 

```
#0 7833.1 ERROR: It looks like you're trying to use Nokogiri as a precompiled native gem on a system with glibc < 2.17:
#0 7833.1 
#0 7833.1   /lib64/libm.so.6: version `GLIBC_2.29' not found (required by /usr/share/gems/gems/nokogiri-1.13.6-aarch64-linux/lib/nokogiri/2.7/nokogiri.so) - /usr/share/gems/gems/nokogiri-1.13.6-aarch64-linux/lib/nokogiri/2.7/nokogiri.so
#0 7833.1 
#0 7833.1   If that's the case, then please install Nokogiri via the `ruby` platform gem:
#0 7833.1       gem install nokogiri --platform=ruby
#0 7833.1   or:
#0 7833.1       bundle config set force_ruby_platform true
#0 7833.1 
#0 7833.1   Please visit https://nokogiri.org/tutorials/installing_nokogiri.html for more help.
#0 7833.1 
#0 7833.1 rake aborted!
#0 7833.1 LoadError: /lib64/libm.so.6: version `GLIBC_2.29' not found (required by /usr/share/gems/gems/nokogiri-1.13.6-aarch64-linux/lib/nokogiri/2.7/nokogiri.so) - /usr/share/gems/gems/nokogiri-1.13.6-aarch64-linux/lib/nokogiri/2.7/nokogiri.so
```

**Review by aebruno — CHANGES_REQUESTED** [review]
@johrstrom Ran into a few arm64 build issues with this PR. The necessary fixes I had to make to get the arm64 builds working are detailed below. If it's OK with you I may close this PR and submit a new one with your changes merged in these minor tweaks. I also plan on adding in the `nohttpd` fixes for vnc.

**Pull request closed without merging**